feat(ts2dart): support property parameter shorthand.

alexeagle opened the pull request:
This lets us declare properties in the constructor parameters in TypeScript code,
and emits explicit property declarations and the "this.prop" constructor param
which is the equivalent Dart shorthand.

Fixes #15

alexeagle committed:
feat(ts2dart): support property parameter shorthand.

This lets us declare properties in the constructor parameters in TypeScript code,
and emits explicit property declarations and the "this.prop" constructor param
which is the equivalent Dart shorthand.

Fixes #15

mprobst commented on lib/declaration.ts:
```diff
@@ -82,12 +82,27 @@ class DeclarationTranspiler extends base.TranspilerStep {
+        let normalizedParameters: ts.ParameterDeclaration[] = [];
+        // Synthesize explicit properties for 'property parameters'
+        ctorDecl.parameters.forEach((param) => {
+            if (this.hasFlag(param.modifiers, ts.NodeFlags.Public) ||
```
indent

mprobst commented on lib/declaration.ts:
```diff
@@ -246,7 +261,7 @@ class DeclarationTranspiler extends base.TranspilerStep {
   }
 
-  private visitParameters(fn: ts.FunctionLikeDeclaration) {
+  private visitParameters(fn: {parameters: ts.ParameterDeclaration[]}) {
```
Maybe just take parameters?

mprobst commented on lib/declaration.ts:
```diff
@@ -272,6 +287,26 @@ class DeclarationTranspiler extends base.TranspilerStep {
+   */
+  private buildSyntheticThisParam(param: ts.ParameterDeclaration): ts.ParameterDeclaration {
+    // The parse tree is immutable, so we build a copy by re-parsing from source.
+    let source = ts.createSourceFile("", `function f(${param.name.getText()}) {}`, ts2dart.OPTIONS.target, true);
```
single quotes

mprobst commented:
Also can you add an integration test? My understanding was that this is valid TS:

```
class X { constructor(private y) {} }
```

But it'd have to translate to this Dart code:

```
class X {
  y;
  X(this.y) {}
}
```

I.e. in Dart you have to declare the field, but can then more easily initialize it. This code doesn't have the code to declare the field, does it?

alexeagle commented:
Added the integration test, and fixed up some of the AST re-write we talked about.
This snapshot still has the synthetic node to let us share the logic for writing out a property declaration, but I suspect we could remove that too. Will look again when I get to my desk.

mprobst commented on lib/declaration.ts:
```diff
@@ -273,6 +268,30 @@ class DeclarationTranspiler extends base.TranspilerStep {
+   */
+  private visitProperty(propertyDecl: ts.PropertyDeclaration | ts.ParameterDeclaration,
+                        containingClass: base.ClassLike) {
+    var hasConstCtor = this.isConst(containingClass);
```
inline?

mprobst commented on test/e2e/lib.ts:
```diff
@@ -1,15 +1,13 @@
-  private _field: string;
-
-  constructor(someVal: string) { this._field = someVal; }
+  constructor(private someVal: string) {}
```
please make sure we have some code in e2e that uses a field

alexeagle replied:
fixed

mprobst commented:
Looks very nice indeed!

alexeagle commented:
merged as d12a7b1131d64ae95d48f9248ec10d44d2db1314
